plugins/connection/grpc.py

```diff
@@ -0,0 +1,294 @@
+        channel, if the value is None it reads the root certificates from
+        a default location chosen by gRPC at runtime.
+    ini:
+      - section: grpc_connection
+        key: root_certificates_file
+    env:
+      - name: ANSIBLE_ROOT_CERTIFICATES_FILE
+    vars:
+      - name: ansible_root_certificates_file
+  certificate_chain_file:
+    description:
+      - The PEM encoded certificate chain file used to create a SSL-enabled
+        channel. If the value is None, no certificate chain is used.
+    ini:
+      - section: grpc_connection
+        key: certificate_chain_file
+    env:
+      - name: ANSIBLE_CERTIFICATE_CHAIN_FILE
+    vars:
+      - name: ansible_certificate_chain_file
+  ssl_target_name_override:
+    description:
+      - The option overrides SSL target name used for SSL host name checking.
+        The name used for SSL host name checking will be the target parameter
+        (assuming that the secure channel is an SSL channel). If this parameter is
+        specified and the underlying is not an SSL channel, it will just be ignored.
+    ini:
+      - section: grpc_connection
+        key: ssl_target_name_override
+    env:
+      - name: ANSIBLE_GPRC_SSL_TARGET_NAME_OVERRIDE
+    vars:
+      - name: ansible_grpc_ssl_target_name_override
+  grpc_type:
+    description:
+        - This option indicates the grpc type and it can be used
+          in place of network_os.
+    default: False
+    ini:
+      - section: grpc_connection
+        key: type
+    env:
+      - name: ANSIBLE_GRPC_CONNECTION_TYPE
+    vars:
+      - name: ansible_grpc_connection_type
+"""
+
+from importlib import import_module
+
+from ansible.errors import AnsibleConnectionFailure, AnsibleError
+from ansible.plugins.connection import NetworkConnectionBase
+
+try:
+    from grpc import insecure_channel, secure_channel, ssl_channel_credentials
+    from grpc.beta import implementations
+
+    HAS_GRPC = True
+except ImportError:
+    HAS_GRPC = False
+
+try:
+    from google import protobuf  # noqa: F401
+
+    HAS_PROTOBUF = True
+except ImportError:
+    HAS_PROTOBUF = False
+
+
+class Connection(NetworkConnectionBase):
+    """GRPC connections"""
+
+    transport = "ansible.netcommon.grpc"
+    has_pipelining = False
+
+    def __init__(self, play_context, new_stdin, *args, **kwargs):
+
+        super(Connection, self).__init__(
+            play_context, new_stdin, *args, **kwargs
+        )
+
+        # TODO: Need to add support to make grpc connection work with non-network target host.
+        # Currently this works only with network target host.
+        grpc_type = self._network_os or self.get_option("grpc_type")
+        if grpc_type:
+            if not HAS_PROTOBUF:
+                raise AnsibleError(
+                    "protobuf is required to use the grpc connection type. Please run 'pip install protobuf'"
+                )
+            if self._network_os:
```

Comment:
replace `self._network_os` with `grpc_type` wherever applicable

Reply:
This if condition is for the case when grpc_type is not given and network_os is given. From the network_os, grpc_type will be formed.

Reply:
`grpc_type` will be FQCN right? In that case the `if else` clause is not required. IIUC the user can set either `grpc_type` or `ansible_network_os` to `cisco.iosxr.grpc`

Reply:
Is it not expected from the user to set ansible_network_os as 'cisco.iosxr.iosxr' ?